packages/ic-asset-certification/src/asset_router.rs

```diff
@@ -886,6 +1099,177 @@ mod tests {
+            assert_matches!(
+                witness.lookup_subtree(&expr_path),
+                SubtreeLookupResult::Found(_)
+            );
+            add_v2_certificate_header(
+                &data_certificate(),
+                &mut expected_response,
+                &witness,
+                &expr_path,
+            );
+            assert_eq!(response, expected_response);
+            asset_len_so_far += response.body().len();
+            number_of_chunks_so_far += 1;
+        }
+        assert_eq!(number_of_chunks_so_far, expected_number_of_chunks)
+    }
+
+    #[rstest]
+    #[case("/long_asset_two_chunks", TWO_CHUNKS_ASSET_LEN)]
+    #[case("/long_asset_six_chunks", SIX_CHUNKS_ASSET_LEN)]
+    #[case("/long_asset_ten_chunks", TEN_CHUNKS_ASSET_LEN)]
+    fn test_long_asset_deletion_removes_chunks(
+        mut long_asset_router: AssetRouter,
+        #[case] req_url: &str,
+        #[case] asset_len: usize,
+    ) {
+        let mut all_requests = vec![];
+        // Request the entire asset and the chunks, all should succeed.
+        // First the asset...
+        let request = HttpRequest::get(req_url).build();
+        let response = long_asset_router
+            .serve_asset(&data_certificate(), &request)
+            .unwrap();
+        let (witness, expr_path) = extract_witness_expr_path(&response);
+
+        assert_eq!(expr_path, vec!["http_expr", &req_url[1..], "<$>"]);
+        assert_matches!(
+            witness.lookup_subtree(&expr_path),
+            SubtreeLookupResult::Found(_)
+        );
+        assert_eq!(response.status_code(), 206);
+        all_requests.push(request);
+
+        // ... then the subsequent chunks.
+        let expected_number_of_chunks =
+            (asset_len as f32 / ASSET_CHUNK_SIZE as f32).ceil() as usize;
+        let mut asset_len_so_far = response.body().len();
+        let mut number_of_chunks_so_far = 1;
+        while asset_len_so_far < asset_len {
+            let chunk_request = HttpRequest::get(req_url)
+                .with_headers(vec![(
+                    "range".to_string(),
+                    format!("bytes={}-", asset_len_so_far),
+                )])
+                .build();
+            let response = long_asset_router
+                .serve_asset(&data_certificate(), &chunk_request)
+                .unwrap();
+            let (witness, expr_path) = extract_witness_expr_path(&response);
+            assert_matches!(
+                witness.lookup_subtree(&expr_path),
+                SubtreeLookupResult::Found(_)
+            );
+            assert_eq!(response.status_code(), 206);
+            asset_len_so_far += response.body().len();
+            number_of_chunks_so_far += 1;
+            all_requests.push(chunk_request);
+        }
+        assert_eq!(number_of_chunks_so_far, expected_number_of_chunks);
+        assert_eq!(all_requests.len(), expected_number_of_chunks);
+
+        // Delete the asset.
```

Comment:
I think it could also be merged into the previous test since they are similar up to this point, what do you think?

Reply:
This is fine for me 👍